File: packages/sdk/src/install/native.ts
import { createRequire } from 'node:module';
import type { Permissions } from '@internals/schemas';

const esmRequire = createRequire(import.meta.url);

interface NativeBindings {
  verifyIntegrity(data: Buffer, expectedIntegrity: string): string;
  extractTarball(data: Buffer, dest: string): string[];
  resolveVersion(available: string[], range: string): string | null;
  isPathAllowed(requestedPath: string, allowedPaths: string[]): boolean;
  isDomainAllowed(domain: string, allowedDomains: string[]): boolean;
  checkPermissionBudget(budget: NativePermissions, skillPerms: NativePermissions, skillName: string): void;
}

interface NativePermissions {
  network?: { outbound?: string[] } | null;
  filesystem?: { read?: string[]; write?: string[] } | null;
  subprocess?: boolean | null;
}

let _native: NativeBindings | null = null;
let _nativeChecked = false;

function tryLoadNative(): NativeBindings | null {
  if (_nativeChecked) return _native;
  _nativeChecked = true;

  try {
    _native = esmRequire('@tankpkg/sdk-core') as NativeBindings;
  } catch {
    _native = null;
  }
  return _native;
}

export function hasNativeAcceleration(): boolean {
  return tryLoadNative() !== null;
}

export function nativeVerifyIntegrity(data: Buffer, expectedIntegrity: string): string | null {
  const n = tryLoadNative();
  if (!n) return null;
  return n.verifyIntegrity(data, expectedIntegrity);
}

export function nativeExtractTarball(data: Buffer, dest: string): string[] | null {
  const n = tryLoadNative();
  if (!n) return null;
  return n.extractTarball(data, dest);
}

export function nativeResolveVersion(
  available: string[],
  range: string
): { used: false } | { used: true; result: string | null } {
  const n = tryLoadNative();
  if (!n) return { used: false };
  return { used: true, result: n.resolveVersion(available, range) };
}

export function nativeIsPathAllowed(requestedPath: string, allowedPaths: string[]): boolean | null {
  const n = tryLoadNative();
  if (!n) return null;
  return n.isPathAllowed(requestedPath, allowedPaths);
}

export function nativeIsDomainAllowed(domain: string, allowedDomains: string[]): boolean | null {
  const n = tryLoadNative();
  if (!n) return null;
  return n.isDomainAllowed(domain, allowedDomains);
}

function toNativePerms(perms: Permissions): NativePermissions {
  return {
    network: perms.network ? { outbound: perms.network.outbound } : null,
    filesystem: perms.filesystem ? { read: perms.filesystem.read, write: perms.filesystem.write } : null,
    subprocess: perms.subprocess ?? null
  };
}

export function nativeCheckPermissionBudget(
  budget: Permissions,
  skillPerms: Permissions,
  skillName: string
): { used: false } | { used: true; error: string | null } {
  const n = tryLoadNative();
  if (!n) return { used: false };
  try {
    n.checkPermissionBudget(toNativePerms(budget), toNativePerms(skillPerms), skillName);
    return { used: true, error: null };
  } catch (err) {
    return { used: true, error: err instanceof Error ? err.message : String(err) };
  }
}

File: packages/sdk-python/tankpkg/_native.py
from __future__ import annotations

from typing import Any

_native: Any = None
_checked = False


def _try_load() -> Any:
    global _native, _checked
    if _checked:
        return _native
    _checked = True
    try:
        import tankpkg_core
    except ImportError:
        _native = None
    else:
        _native = tankpkg_core
    return _native


def has_native() -> bool:
    return _try_load() is not None


def native_verify_integrity(data: bytes, expected_integrity: str) -> str | None:
    n = _try_load()
    if n is None:
        return None
    return n.verify_integrity(data, expected_integrity)


def native_extract_tarball(data: bytes, dest: str) -> list[str] | None:
    n = _try_load()
    if n is None:
        return None
    return n.extract_tarball(data, dest)


def native_resolve_version(available: list[str], range_: str) -> tuple[bool, str | None]:
    n = _try_load()
    if n is None:
        return (False, None)
    return (True, n.resolve_version(available, range_))


def native_is_path_allowed(requested_path: str, allowed_paths: list[str]) -> bool | None:
    n = _try_load()
    if n is None:
        return None
    return n.is_path_allowed(requested_path, allowed_paths)


def native_is_domain_allowed(domain: str, allowed_domains: list[str]) -> bool | None:
    n = _try_load()
    if n is None:
        return None
    return n.is_domain_allowed(domain, allowed_domains)


def _reset_for_tests() -> None:
    global _native, _checked
    _native = None
    _checked = False
